core/domain/stats_jobs_one_off.py

```diff
@@ -1694,3 +1694,159 @@ def reduce(key, values):
+            new_exp_version = None
+
+            if revert_to_version is not None:
+                if revert_to_version == 1:
+                    yield (
+                        FillExplorationStatsOneOffJob.V1_EXPLORATION_MODEL,
+                        exp.id)
+                    return
+                exp_versions_diff = None
+                prev_exp_stats = exp_stats_list[revert_to_version - 2]
+                prev_exp = exp_list[revert_to_version - 2]
+                new_exp_version = revert_to_version
+            else:
+                if exp.version == 1:
+                    yield (
+                        FillExplorationStatsOneOffJob.V1_EXPLORATION_MODEL,
+                        exp.id)
+                    return
+                exp_versions_diff = exp_domain.ExplorationVersionsDiff(
+                    change_list)
+                prev_exp_stats = exp_stats_list[exp.version - 2]
+                prev_exp = exp_list[exp.version - 2]
+                new_exp_version = exp.version
+
+            # Fill missing Exploration-level stats.
+            if exp_stats is not None:
+                num_valid_exp_stats += 1
+            else:
+                exp_stats = prev_exp_stats and prev_exp_stats.clone()
+
+                if exp_versions_diff is not None:
+                    exp_stats = stats_services.advance_version_of_exp_stats(
+                        new_exp_version, exp_versions_diff, exp_stats, None,
+                        None)
+                else:
+                    exp_stats.version = exp.version
+                exp_stats_list[i] = exp_stats
+                missing_exp_stats.append(
+                    'Exp id: %s, Version: %s' % (exp_id, exp.version))
+                new_exploration_stats = (
+                    stats_services.get_stats_for_new_exploration(
+                        exp_id, exp.version, exp.states))
+                stats_services.create_stats_model(new_exploration_stats)
+
+            # Fill missing State-level stats.
+            state_stats_mapping = exp_stats.state_stats_mapping
+            for state_name, _ in exp.states.items():
+                if state_name in state_stats_mapping:
+                    num_valid_state_stats += 1
+                    continue
+
+                if exp_versions_diff is not None:
+                    prev_state_name = (
+                        exp_versions_diff.new_to_old_state_names.get(
+                            state_name, state_name))
+                else:
+                    prev_state_name = state_name
+
+                state_stats_mapping[state_name] = (
+                    prev_exp_stats.state_stats_mapping[prev_state_name].clone()
+                    if exp.interaction.id == prev_exp.interaction.id else
+                    stats_domain.StateStats.create_default())
+                missing_state_stats.append(
+                    'Exp id: %s, Version: %s, State name: %s' % (
```

Comment:
```suggestion
                    'StateStats(exp_id=%r, exp_version=%r, state_name=%r)' % (
```

nit: I think it'd be more readable/helpful for the job output to use Pythonic string outputs.

Reply:
Done.